### ISSUE TITLE
Enforce migrations to be standalone scripts

### ISSUE THREAD

**Issue by kianenigma** [paritytech/substrate#10308]
Follow up on the discussion here: https://github.com/paritytech/substrate/pull/10233#discussion_r749441704. Currently, we keep some migration codes in the code, and even those that we do are pretty much useless. Given this our support for historical migrations are basically near-nothing. 

I propose we enforce all migrations to be self-containing henceforth. This means that they should not depend on anything from the pallet. 

If they need a struct, they must duplicate that version of it in their own local scope. If they need a particular storage item, they can mimic it using `generate_storage_alias`.

Next to this, we are moving toward banning `on_runtime_upgrade` inside the pallet. This is aligned with this issue as well. To begin with, migrations should be controlled explicitly by the top-level runtime.  

Once we have all of this, I belive we can make migrations fully automated. Each migration will have a u32 id. Each pallet will have an on-chain u32 identifier, akin to the current storage version. Upon each runtime upgrade, if any migrations have been added to the repository will be executed. These migrations need to be added to some static registry. 

Regardless of realizing the last statement or not, in order to for the migrations codes that we keep in substrate to have any utility, we better start by making them independent of the pallet. Some examples of this already exist (`pallet-elections-phragmen`).

**Comment by KiChjang** [comment]
Would this make migrations look more or less akin to database migrations for a server backend? I think that should ideally be what we're aiming for here.

**Comment by kianenigma** [comment]
> Would this make migrations look more or less akin to database migrations for a server backend? I think that should ideally be what we're aiming for here.

totally.

**Comment by bkchr** [comment]
> Running prior migrations is not a problem either, just using release branches / webassembly code from older versions. It's a bit of a pain, but I don't see this happening often either.

You are forgetting that Polkadot isn't the only user of Substrate. For staking this currently maybe being true, but most of the other pallets are being used by a lot of Parachains etc. Your complaints are nevertheless correct, but we should work on this. We need to identify common patterns and then provide abstractions/macros/tools to help with that.

**Comment by ggwpez** [comment]
https://github.com/paritytech/substrate/issues/13107 tries to abstract the redundant version tracking and logging.

>In case of replicating, say, BagsList pallet storage and methods we'd have to replicate several pretty intricate methods along with their corresponding structs:

Uhf… yea that is annoying. Having mutation functions on stored structs makes this much more complicated. I dont see how we could circumvent that besides not using that pattern in the first place.

**Comment by ruseinov** [comment]
> #13107 tries to abstract the redundant version tracking and logging.
> 
> > In case of replicating, say, BagsList pallet storage and methods we'd have to replicate several pretty intricate methods along with their corresponding structs:
> 
> Uhf… yea that is annoying. Having mutation functions on stored structs makes this much more complicated. I dont see how we could circumvent that besides not using that pattern in the first place. Which is sub-optimal, since its in itself a good pattern.

If I'm being honest I don't see why the actual library code should suffer for migration convenience. If anything it should be the other way around. Migrations are simply one-off helpers that serve the purpose, once. 

It's a bit annoying having to potentially use several different release branches to gradually upgrade, but it can be done and it's not that hard. Maybe we should provide some sort of convenience tooling for that instead?

**Comment by ruseinov** [comment]
Now that I re-read the original issue it's a bit more clear to me what the purpose of this is: migration automation.

Still, I can't see how this could be easily achieved as decoupling migrations from pallets is tricky enough, but we also still need to specify the correct storage version of each pallet should anyone start from the current codebase. And if this is wrong and somehow slips into master - that'll be potentially a huge pain as things would fail while trying to roll those migrations out automatically.

It seems like manual control over migrations similar to what we have now, having to add them to individual runtimes, might be a better option. It's a bit more work, but it gives more fine-grained control of what's happening, since we need to compile a new runtime anyway in order to introduce any new code. 

I've had similar experiences with SQL migrations on various projects. If the migration is complex enough - it might be a multi-step one, that needs to be run and monitored manually. Even if it's not that complex - it's almost always a dangerous pattern to let it be run automatically or via CI. In our case the automation would come from simply compiling a new runtime with new migrations in place.
One other concern is: what if we introduce 10 migrations to 10 different pallets, if all of those run back-to-back - that could potentially be a concern.

All of the above concerns might be good to consider in order to see how we could mitigate some if not all of them while still providing a reasonable level of automation.

**Comment by kianenigma** [comment]
I think we should not settle this yet and as Basti said try and find patters that are common and help migration scripts live longer. One of them is this: 


https://github.com/paritytech/substrate/blob/0246883c404d498090f33e795feb8075fa8d3b6b/frame/elections-phragmen/src/migrations/v3.rs#L41-L48

Which means: 

1. don't rely on the definition of storage structs from your pallet
2. don't rely on any type coming in from `trait Config` of your pallet

Using, your PR which sparked this discussion can be solved fairly easily https://github.com/paritytech/substrate/pull/13079/commits/2c6275e91d188ccc236ec85760f05ba0d3aff01a as the only breaking change is the introduction of `ReadOnlySortedListProvider`.

That being said, this is still not perfect because i.e. if the definition of `SortedListProvider` changes and one of the functions used in the migrations are i.e. removed, then we would again be trouble. That being said, again, that is also solvable. 


TLDR; I don't have a plan here, but I am not at a spot where I can also say screw the old migrations and let's move on. If Substrate wants to be ultimate, user-friendly framework that it wants to be, it must solve this issue, and I am down to put more effort into it.

**Comment by athei** [comment]
Let us take a step back and think what would be the most convenient and easiest way to deliver migrations for our users:

I think it is the old way of bundling them with the pallet. The pallet knows its current version and the version in storage. It runs all necessary migrations. The runtime just calls into all pallets for migrations on upgrade. Everything just works.

But there are reasons why we are no longer doing it this way. Let me see if I got all of them:

- Migrations consume space in the runtimes code. We don't want them in the pallets and hence in the productive runtime.
- Migrations are more opaque to runtime authors when handled by the pallets. This is nice but this might lead to a situation where they compound to something that is heavier than a single block.

I think we kind find ways to address both issues without dumping this hugely complicated topic onto our users.

**Comment by ruseinov** [comment]
> The idea that I am now eager to implement and show to all of you as PoC is this: I will migrate all of the migrations of the nomination pools pallet to be 100% independent of the runtime in some PR, and we can then see what the outcome looks like. If someone else has the time, please go ahead.

This is not documented in the issue, but I think @kianenigma is not that big of a fan of this idea anymore, as discussed during FRAME steering meet. Or was it Staking meet? I don't remember. Anyhow, since we don't have a unified way/language (SQL-like) to deal with this and a lot of the migrations really need to mimic the business-logic of pallets whose storage they are touching - it seems like the best approach is to do what we've done before. 



> * Migrations consume space in the runtime's code. We don't want them in the pallets and hence in the productive runtime.

It seems like a good idea is to clean up old migrations. And let the users deal with this by gradually upgrading, if they need to. But yeah, the UX of this is not amazing.
